### ISSUE TITLE
Chart filtering for allmetrics endpoint (json and shell formats)

### ISSUE THREAD

**Issue by mahlonsmith** [netdata/netdata#3198]
Hey Costa!

I have a couple of enhancement requests for the `[backend]` section of the netdata configuration.

Currently, the "send charts matching" attribute nicely filters collected data, but it only does so for the prometheus format.  How difficult would it be to allow that filtering to function for all formats supported by the *allmetrics* API endpoint?

Secondarily (or perhaps, alternatively):  That endpoint allows overriding server configuration of specific attributes (names, prefix).  It would be exceptionally nice to expose the `send charts matching` behavior as a parameter to the allmetrics endpoint as well.

I realize I can fetch directly at the `chart` endpoint, but that only works when I know the chart name in advance -- there's quite a few that are dynamic, that necessitate the allmetrics fetching -- only to throw away all the rest of the data that I'm not interested in -- roughly 1MB of JSON, * N servers, * X requests/min... it adds up really fast.

Something like:

    http://example.org:19999/api/v1/allmetrics?format=json&filter=disk_space.*%20system.load

Thanks for any feedback!

**Comment by ktsaou** [comment]
Hi,

Well, I made a PR to allow `allmetrics?format=json` to use `send charts matching`, but I think this is wrong and we should not merge it as-is.

Let me explain.

`allmetrics?format=json` allows netdata to expose the **latest values** of all charts in json format. It is to be used by front-end programs willing to provide some kind of different view of the current system state. Think of it as the backend for a program like remote `top`, remote `vmstat`, etc.

Backends have a different feature set. For example the prometheus backend respects the data source (`as-collected`, `average`, `sum`), and when it is `average` or `sum` it remembers the last time it was accessed by each client and provides the average of all the data since the last call (so that prometheus gets an average of the whole time, without gaps). `allmetrics?format=json` does not have these features.

I am also somewhat skeptical to add `send charts matching` to the URL. The way it is now, netdata caches this decision. It does not re-evaluate the pattern matches on every call. netdata does 2 matches per chart (chart id and chart name) and there are several hundreds of charts per host. We can easily end up with 1000 pattern matches per call per host (imagine the number of matches on a master host mirroring 30 other hosts - it will have to do 30000 matches per API call).

I understand the confusion though. Probably we should:

1. rename `allmetrics?format=prometheus` to something else (`backend_pull?format=prometheus` ?), to avoid this confusion.

2. provide another backend `backend_pull?format=json` that will have all the features of the backends, similar to prometheus.

3. add `filter=PATTERN` to both `backend_pull` and `allmetrics` and find a way to index it and re-use it across multiple calls.

Keep also in mind there a `json` backend that has all the features of the backends and works the other way around: netdata pushes metrics to the backend.

**Comment by mahlonsmith** [comment]
This all makes good sense -- sounds like the larger issue is just that the behavioral differences for a single format overload (or at least muddy) the real intent of the endpoint.  Thank you for clearing that up.

> Keep also in mind there is a json backend that has all the features of the backends and works the other way around: netdata pushes metrics to the backend.

I completely missed that as a backend format option.  Neat!  That may come in handy for some alternative stuff later, but for this round I'll definitely be strobing netdata clients vs having a central backstop -- separate timers gathering *specific* data at differing intervals, based on need (load every minute, disk every 60 mins, etc), with a focus on minimizing network load at scale.  Lots of ways to accomplish that with aggregation and outside tools, but there are some internal limitations/motivations for keeping it as simplistic as possible... ie, do it all in netdata requests without involving other systems.  (I'm happy to discuss details offline more if you'd like!)

In any event, all of the above sounds like it would address exactly what I was after while retaining desired endpoint purposes -- and at a minimum provide people with more options for consuming gathered netdata information! (wooo)  :+1:

**Comment by fayak** [comment]
Actually the `filter` option could be very useful to us, we're using prometheus to query the data from netdata, and we'd like to have a specific prometheus instance for a specific metric (collected amongst others).
I feel like something like this could be very useful:
```
> GET /api/v1/allmetrics?format=prometheus
netdata_cgroup_cpu_limit_percentage_average{chart="cgroup_engage_voice_instance2.cpu_limit",family="cpu",dimension="used"} 0.2641549 1612545474000
netdata_services_throttle_io_ops_write_operations_persec_average{chart="services.throttle_io_ops_write",family="disk",dimension="irqbalance"} 0.0000000 1612545474000
...
> GET /api/v1/allmetrics?format=prometheus&filter=^netdata_cgroup
netdata_cgroup_cpu_limit_percentage_average{chart="cgroup_engage_voice_instance2.cpu_limit",family="cpu",dimension="used"} 0.2641549 1612545474000
> GET /api/v1/allmetrics?format=prometheus&filter=^netdata_cgroup&filter_negate=true
netdata_services_throttle_io_ops_write_operations_persec_average{chart="services.throttle_io_ops_write",family="disk",dimension="irqbalance"} 0.0000000 1612545474000
...
```

Being able to use regexp on the filter, and having an option to invert the query (like the `-v` on `grep`) would be nice, but I have no clues on the work that should be put to make it work though

**Comment by ilyam8** [comment]
Closing this feature request. If you are still interested in it, consider creating a topic on [our community forum](https://community.netdata.cloud/c/feature-requests/agent-fr/21).

**Comment by bioshazard** [comment]
Found an article covering how to drop metrics by regex. Planning to drop any metric other than my whitelisted set to achieve this outcome until it becomes supported natively. https://www.robustperception.io/dropping-metrics-at-scrape-time-with-prometheus

**Comment by ktsaou** [comment]
Thank you @bioshazard !

@ilyam8 @stelfrag @cpipilas since this is a much needed feature, I suggest to reconsider and implement it. We could also support simple patterns at the URL.

**Comment by ilyam8** [comment]
Implemented in #12566
